src/test/java/com/google/cloud/spanner/pgadapter/python/Django/DjangoBasicTest.java

```diff
@@ -0,0 +1,191 @@
+            .setRowType(
+                StructType.newBuilder()
+                    .addFields(
+                        Field.newBuilder()
+                            .setType(Type.newBuilder().setCode(TypeCode.INT64).build())
+                            .setName("singerid")
+                            .build())
+                    .addFields(
+                        Field.newBuilder()
+                            .setType(Type.newBuilder().setCode(TypeCode.STRING).build())
+                            .setName("firstname")
+                            .build())
+                    .addFields(
+                        Field.newBuilder()
+                            .setType(Type.newBuilder().setCode(TypeCode.STRING).build())
+                            .setName("lastname")
+                            .build())
+                    .build())
+            .build());
+    for (int i = 0; i < rows.size(); i += 3) {
+      String singerid = rows.get(i), firstname = rows.get(i + 1), lastname = rows.get(i + 2);
+      resultSetBuilder.addRows(
+          ListValue.newBuilder()
+              .addValues(Value.newBuilder().setStringValue(singerid).build())
+              .addValues(Value.newBuilder().setStringValue(firstname).build())
+              .addValues(Value.newBuilder().setStringValue(lastname).build())
+              .build());
+    }
+    return resultSetBuilder.build();
+  }
+
+  @Test
+  public void testSelectAll() throws IOException, InterruptedException {
+    String sqlSelectAll =
+        "SELECT \"singers\".\"singerid\", \"singers\".\"firstname\", \"singers\".\"lastname\" FROM \"singers\"";
+
+    List<String> result =
+        new ArrayList<>(Arrays.asList("1", "hello", "world", "2", "hello", "django"));
+
+    mockSpanner.putStatementResult(
+        StatementResult.query(Statement.of(sqlSelectAll), createResultSet(result)));
+    String expectedOutput =
+        "{'singerid': 1, 'firstname': 'hello', 'lastname': 'world'}\n"
+            + "{'singerid': 2, 'firstname': 'hello', 'lastname': 'django'}\n";
+    List<String> options = new ArrayList<>(Arrays.asList("all"));
+    String actualOutput = executeBasicTests(pgServer.getLocalPort(), host, options);
+    assertEquals(expectedOutput, actualOutput);
+
+    assertEquals(1, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+    assertEquals(
+        sqlSelectAll, mockSpanner.getRequestsOfType(ExecuteSqlRequest.class).get(0).getSql());
+  }
+
+  @Test
+  public void testInsert() throws IOException, InterruptedException {
+    String sqlUpdate =
+        "UPDATE \"singers\" SET \"firstname\" = 'john', \"lastname\" = 'doe' WHERE \"singers\".\"singerid\" = 4";
+    String sqlInsert =
+        "INSERT INTO \"singers\" (\"singerid\", \"firstname\", \"lastname\") VALUES (4, 'john', 'doe')";
+
+    // Django first sends an update statement
+    // If the updated_rowcount returned by the server is 0, then only it sends the insert statement
```

Comment:
Is this the default behavior of Django, or do you need to configure something specific to get this behavior? The reason that I ask is that PostgreSQL supports [insert-or-update](https://www.postgresql.org/docs/15/sql-merge.html) statements, and I would have expected Django to use that by default instead of two separate DML statements.

Reply:
Yeah, it is the default behaviour of the Django and I was surprised too. There are some fields like `force_insert` and `force_update` in the `save` function to enforce the insert or update but by default, it works like this only.